datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java

```diff
@@ -117,7 +117,7 @@ public LoadBasedPartitionAssignmentStrategy(PartitionThroughputProvider throughp
     // Doing assignment
     LoadBasedPartitionAssigner partitionAssigner = new LoadBasedPartitionAssigner();
     return partitionAssigner.assignPartitions(clusterThroughputInfo, currentAssignment,
-        unassignedPartitions, datastreamPartitions);
+        unassignedPartitions, datastreamPartitions, _partitionsPerTask);
```

Comment:
`_partitionsPerTask` represents the ideal number of partition distribution per task which is used along with `_partitionFullnessFactorPct` which represent how much percent to fill. This count is not a hard check, but a desired count. `_maxPartitionsPerTask` is the ultimate count is enforced. If you enforce based on `_partitionsPerTask` and `numTasks` are getting limited by smaller `maxTask` configuration, the entire pipeline will be stuck till the `maxTasks` configuration is not corrected. One of the most important purpose of enforcing `_maxPartitionsPerTask` is to avoid zk node size going > 1MB.

Reply:
To add to above point, If `_enablePartitionNumBasedTaskCountEstimation` is enabled, then we can consider honoring `_partitionsPerTask` and `_partitionFullnessFactorPct` along with `_maxPartitionsPerTask`, which means we will first try to fill in upto `_partitionsPerTask` if we can, and if we cannot then will stretch upto `_maxPartitionsPerTask`.

There is a scope of reusing the alert here `PARTITIONS_PER_TASK_NEEDS_ADJUSTMENT`

Reply:
At the moment, I'm simply using `_maxPartitionsPerTask`. We can discuss usage of `partitionsPerTask` and `partitionFullnessFactorPct` offline.